server/sonar-server/src/main/java/org/sonar/server/measure/ws/ComponentTreeAction.java

```diff
@@ -0,0 +1,355 @@
+      .setRequired(true)
+      .setExampleValue("ncloc, complexity");
+
+    action.createParam(PARAM_METRIC_SORT)
+      //TODO description // NOSONAR
+      .setDescription("Metric key to sort by. Make sure it's in the list of metric keys")
+      .setExampleValue("ncloc");
+
+    action.createParam(PARAM_ADDITIONAL_FIELDS)
+      //TODO add description // NOSONAR
+      .setDescription("Additional field in the response")
+      .setPossibleValues("metrics");
+
+    createQualifiersParameter(action, newQualifierParameterContext(userSession, i18n, resourceTypes));
+
+    action.createParam(PARAM_STRATEGY)
+      .setDescription("Strategy to search for base component children:" +
+        "<ul>" +
+        "<li>children: return the direct children components of the base component. Grandchildren components are not returned</li>" +
+        "<li>all: return all the children components of the base component. Grandchildren are returned.</li>" +
+        "<li>leaves: return all the children components (files, in general) which don't have other children. They are the leaves of the component tree.</li>" +
+        "</ul>")
+      .setPossibleValues(STRATEGIES)
+      .setDefaultValue(ALL_STRATEGY);
+  }
+
+  @Override
+  public void handle(Request request, Response response) throws Exception {
+    ComponentTreeWsResponse componentTreeWsResponse = doHandle(toComponentTreeWsRequest(request));
+    writeProtobuf(componentTreeWsResponse, request, response);
+  }
+
+  private ComponentTreeWsResponse doHandle(ComponentTreeWsRequest request) {
+    DbSession dbSession = dbClient.openSession(false);
+    try {
+      ComponentDto baseComponent = componentFinder.getByUuidOrKey(dbSession, request.getBaseComponentId(), request.getBaseComponentKey(), BASE_COMPONENT_ID_AND_KEY);
+      checkPermissions(baseComponent);
+      SnapshotDto baseSnapshot = dbClient.snapshotDao().selectLastSnapshotByComponentId(dbSession, baseComponent.getId());
+      if (baseSnapshot == null) {
+        return emptyResponse(request);
+      }
+
+      ComponentTreeQuery query = toComponentTreeQuery(request, baseSnapshot);
+      ComponentDtosAndTotal componentDtosAndTotal = searchComponents(dbSession, query, request.getStrategy());
+      List<MetricDto> metrics = searchMetrics(dbSession, request);
+      Table<String, MetricDto, MeasureDto> measuresByComponentUuid = searchMeasuresByComponentUuidAndMetric(dbSession, componentDtosAndTotal.componentDtos, metrics);
+
+      return buildResponse(
+        componentDtosAndTotal.componentDtos,
+        metrics,
+        measuresByComponentUuid,
+        Paging.forPageIndex(query.getPage()).withPageSize(query.getPageSize()).andTotal(componentDtosAndTotal.total));
+    } finally {
+      dbClient.closeSession(dbSession);
+    }
+  }
+
+  private static ComponentTreeWsResponse buildResponse(List<ComponentDtoWithSnapshotId> components, List<MetricDto> metricDtos, Table<String, MetricDto, MeasureDto> measuresByComponentUuid, Paging paging) {
```

Comment:
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png) Split this 206 characters long line (which is greater than 180 authorized). [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS00103)